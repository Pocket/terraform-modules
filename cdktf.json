{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
