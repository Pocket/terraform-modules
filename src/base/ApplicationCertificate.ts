import { AcmCertificate } from '@cdktf/provider-aws/lib/acm-certificate';
import { AcmCertificateValidation } from '@cdktf/provider-aws/lib/acm-certificate-validation';
import { DataAwsRoute53Zone } from '@cdktf/provider-aws/lib/data-aws-route53-zone';
import { Route53Record } from '@cdktf/provider-aws/lib/route53-record';
import { TerraformMetaArguments, TerraformProvider } from 'cdktf';
import { Construct } from 'constructs';

export interface ApplicationCertificateProps extends TerraformMetaArguments {
  domain: string;
  /**
   * If zoneId is not passed then we use a data block and the zoneDomain to grab it.
   */
  zoneId?: string;
  zoneDomain?: string;
  tags?: { [key: string]: string };
}

/**
 * Generates an Application Certificate given a domain name and zoneId
 */
export class ApplicationCertificate extends Construct {
  public readonly arn: string;
  // Use `certificateValidation` in `dependsOn` to block on the
  // complete certificate for any downstream dependencies
  public readonly certificateValidation: AcmCertificateValidation;

  constructor(
    scope: Construct,
    name: string,
    config: ApplicationCertificateProps
  ) {
    super(scope, name);

    if (!config.zoneId && config.zoneDomain) {
      const route53Zone = new DataAwsRoute53Zone(this, `zone`, {
        name: config.zoneDomain,
        provider: config.provider,
      });
      config.zoneId = route53Zone.zoneId;
    } else if (!config.zoneId && !config.zoneDomain) {
      throw new Error('You need to pass either a zone id or a zone domain');
    }

    const certificate = ApplicationCertificate.generateAcmCertificate(
      this,
      config.domain,
      config.tags,
      config.provider
    );

    const certificateRecord = ApplicationCertificate.generateRoute53Record(
      this,
      config.zoneId,
      certificate,
      config.provider
    );

    const validation = ApplicationCertificate.generateAcmCertificateValidation(
      this,
      certificate,
      certificateRecord,
      config.provider
    );

    this.arn = certificate.arn;
    this.certificateValidation = validation;
  }

  static generateAcmCertificate(
    scope: Construct,
    domain: string,
    tags?: { [key: string]: string },
    provider?: TerraformProvider
  ): AcmCertificate {
    return new AcmCertificate(scope, `certificate`, {
      domainName: domain,
      validationMethod: 'DNS',
      tags: tags,
      lifecycle: {
        createBeforeDestroy: true,
      },
      provider: provider,
    });
  }

  static generateRoute53Record(
    scope: Construct,
    zoneId: string,
    cert: AcmCertificate,
    provider?: TerraformProvider
  ): Route53Record {
    const record = new Route53Record(scope, `certificate_record`, {
      name: cert.domainValidationOptions.get(0).resourceRecordName,
      type: cert.domainValidationOptions.get(0).resourceRecordType,
      zoneId,
      records: [cert.domainValidationOptions.get(0).resourceRecordValue],
      ttl: 60,
      dependsOn: [cert],
      provider: provider,
    });

    return record;
  }

  static generateAcmCertificateValidation(
    scope: Construct,
    cert: AcmCertificate,
    record: Route53Record,
    provider?: TerraformProvider
  ): AcmCertificateValidation {
    return new AcmCertificateValidation(scope, `certificate_validation`, {
      certificateArn: cert.arn,
      validationRecordFqdns: [record.fqdn],
      dependsOn: [record, cert],
      provider: provider,
    });
  }
}
