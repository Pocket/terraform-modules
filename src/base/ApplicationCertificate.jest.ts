import { Testing, TerraformStack } from 'cdktf';
import { ApplicationCertificate } from './ApplicationCertificate';

test('throws an error without a zone id or zone domain', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  expect(() => {
    new ApplicationCertificate(stack, 'testCert', {
      domain: 'dev.gobowling.info',
    });
  }).toThrow('You need to pass either a zone id or a zone domain');
});

test('renders a cert with a zone id', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'malibu',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders a cert with a zone domain', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'gobowling.info',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders a cert with tags', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'gobowling.info',
    tags: {
      name: 'thedude',
      hobby: 'bowling',
    },
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
