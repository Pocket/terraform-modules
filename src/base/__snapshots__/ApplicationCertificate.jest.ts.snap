// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a cert with a zone domain 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test_testCert_certificate_7873C49D\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"validation_method\\": \\"DNS\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate\\",
            \\"uniqueId\\": \\"test_testCert_certificate_7873C49D\\"
          }
        }
      }
    },
    \\"aws_route53_record\\": {
      \\"test_testCert_certificaterecord_6E221F0E\\": {
        \\"name\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"gobowling.info\\",
        \\"depends_on\\": [
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_record\\",
            \\"uniqueId\\": \\"test_testCert_certificaterecord_6E221F0E\\"
          }
        }
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"test_testCert_certificatevalidation_BAC3E388\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test_testCert_certificate_7873C49D.arn}\\",
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.test_testCert_certificaterecord_6E221F0E.fqdn}\\"
        ],
        \\"depends_on\\": [
          \\"aws_route53_record.test_testCert_certificaterecord_6E221F0E\\",
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_validation\\",
            \\"uniqueId\\": \\"test_testCert_certificatevalidation_BAC3E388\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders a cert with a zone id 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test_testCert_certificate_7873C49D\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"validation_method\\": \\"DNS\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate\\",
            \\"uniqueId\\": \\"test_testCert_certificate_7873C49D\\"
          }
        }
      }
    },
    \\"aws_route53_record\\": {
      \\"test_testCert_certificaterecord_6E221F0E\\": {
        \\"name\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"malibu\\",
        \\"depends_on\\": [
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_record\\",
            \\"uniqueId\\": \\"test_testCert_certificaterecord_6E221F0E\\"
          }
        }
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"test_testCert_certificatevalidation_BAC3E388\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test_testCert_certificate_7873C49D.arn}\\",
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.test_testCert_certificaterecord_6E221F0E.fqdn}\\"
        ],
        \\"depends_on\\": [
          \\"aws_route53_record.test_testCert_certificaterecord_6E221F0E\\",
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_validation\\",
            \\"uniqueId\\": \\"test_testCert_certificatevalidation_BAC3E388\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders a cert with tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test_testCert_certificate_7873C49D\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"tags\\": {
          \\"name\\": \\"thedude\\",
          \\"hobby\\": \\"bowling\\"
        },
        \\"validation_method\\": \\"DNS\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate\\",
            \\"uniqueId\\": \\"test_testCert_certificate_7873C49D\\"
          }
        }
      }
    },
    \\"aws_route53_record\\": {
      \\"test_testCert_certificaterecord_6E221F0E\\": {
        \\"name\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test_testCert_certificate_7873C49D.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"gobowling.info\\",
        \\"depends_on\\": [
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_record\\",
            \\"uniqueId\\": \\"test_testCert_certificaterecord_6E221F0E\\"
          }
        }
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"test_testCert_certificatevalidation_BAC3E388\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test_testCert_certificate_7873C49D.arn}\\",
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.test_testCert_certificaterecord_6E221F0E.fqdn}\\"
        ],
        \\"depends_on\\": [
          \\"aws_route53_record.test_testCert_certificaterecord_6E221F0E\\",
          \\"aws_acm_certificate.test_testCert_certificate_7873C49D\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testCert/certificate_validation\\",
            \\"uniqueId\\": \\"test_testCert_certificatevalidation_BAC3E388\\"
          }
        }
      }
    }
  }
}"
`;
